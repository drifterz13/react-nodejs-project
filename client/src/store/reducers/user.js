export const LOGGED_IN = 'LOGGED_IN'
export const SET_PREF = 'SET_PREF'

const initialState = {
  info: {},
  authenticated: false
}

export default (state = initialState, action) => {
  switch (action.type) {
    case LOGGED_IN:
      return {
        ...state,
        authenticated: true,
        info: { ...action.payload }
      }
    case SET_PREF:
      return {
        ...state,
        info: { ...state.info, preferences: action.payload }
      }
    default:
      return state
  }
}
