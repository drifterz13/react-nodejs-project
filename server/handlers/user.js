require('dotenv').load()
const db = require('../models')
const jwt = require('jsonwebtoken')

module.exports.signup = async function(req, res, next) {
  try {
    const foundUser = await db.User.findOne({ email: req.body.email })
    if (foundUser) {
      return next({
        status: 404,
        message: 'This email is already taken.'
      })
    }
    const newUser = await db.User.create(req.body)
    const { username, email, _id } = newUser
    const token = await jwt.sign({
      username,
      email,
      _id
    }, process.env.JWT_SECRET, {
      expiresIn: 3600
    })
    const responseData = { _id, username, email, token }
    return res.status(200).json(responseData)
  } catch (err) {
    return next(err)  
  }
}

module.exports.signin = async function(req, res, next) {
  try {
    let foundUser = await db.User.findOne({ email: req.body.email })
    if (!foundUser) {
      return next({
        status: 404,
        message: 'Invalid credentials.'
      })
    }
    const isMatch = await foundUser.comparePassword(req.body.password)
    if (foundUser && isMatch) {
      const { username, email, _id, preferences } = foundUser
      const token = await jwt.sign({
        username,
        email,
        _id
      }, process.env.JWT_SECRET, {
        expiresIn: 3600
      })
      const responseData = { _id, username, email, token, preferences }
      return res.status(200).json(responseData)
    } else {
      return next({
        status: 404,
        message: 'Invalid password.'
      })
    }
  } catch (err) {
    return next(err)
  }
}

module.exports.savePreferences = async function(req, res, next) {
  try {
    let foundUser = await db.User.findById(req.params.id)
    const {
      language,
      timezone,
      currency,
      profileType,
      messageType,
      categoryType
    } = req.body
    const preferences = {
      language,
      timezone,
      currency,
      profileType,
      messageType,
      categoryType
    }
    foundUser.preferences = preferences
    foundUser.save()
    return res.status(200).json(preferences)
  } catch (err) {
    return next(err)
  }
}