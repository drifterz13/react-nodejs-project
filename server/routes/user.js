const express = require('express')
const router = express.Router({ mergeParams: true })
const { signup, signin, savePreferences } = require('../handlers/user')

router.post('/signup', signup)
router.post('/signin', signin)
router.route('/:id/pref').put(savePreferences)

module.exports = router