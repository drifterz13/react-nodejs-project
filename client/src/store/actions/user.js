import axios from 'axios'
import { LOGGED_IN, SET_PREF } from '../reducers/user'

export const loggedIn = user => {
  return {
    type: LOGGED_IN,
    payload: user
  }
}

export const settingPref = data => {
  return {
    type: SET_PREF,
    payload: data
  }
}

export const register = credentials => async (dispatch) => {
  try {
    const { data } = await axios.post('http://localhost:5000/auth/signup', credentials)
    console.log(data)
    localStorage.setItem('token', data.token)
    return dispatch(loggedIn(data))
  } catch (err) {
    return Promise.reject(err.response.data.error)
  }
}

export const login = credentials => async (dispatch) => {
  try {
    const { data } = await axios.post('http://localhost:5000/auth/signin', credentials)
    localStorage.setItem('token', data.token)
    return dispatch(loggedIn(data))
  } catch (err) {
    return Promise.reject(err.response.data.error)
  }
}

export const setPreferences = (preferences, id) => {
  return async dispatch => {
    try {
      const { data } = await axios.put(`http://localhost:5000/auth/${id}/pref`, preferences)
      console.log(data)
      dispatch(settingPref(data))
    } catch (err) {
      return Promise.reject(err.response)
    }
  }
}
