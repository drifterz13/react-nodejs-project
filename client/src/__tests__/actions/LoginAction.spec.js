import configureStore from 'redux-mock-store'
import { loggedIn } from '../../store/actions/user'

describe('User login action', () => {
  test('Dispatches the correct action and payload', () => {
    const mockStore = configureStore()
    const store = mockStore()
    store.clearActions()
    const mockData = {
      username: 'test',
      email: 'test@test.com',
      password: 'test'
    }
    store.dispatch(loggedIn(mockData))
    expect(store.getActions()).toMatchSnapshot()
  })
})