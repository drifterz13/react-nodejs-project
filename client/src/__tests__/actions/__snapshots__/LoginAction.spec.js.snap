// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User login action Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": Object {
      "email": "test@test.com",
      "password": "test",
      "username": "test",
    },
    "type": "LOGGED_IN",
  },
]
`;
